File: competitive_programming/leetcode/python/47_permutations_ii.py
def permuteUnique(nums):
  def dfs(nums, stack, used, ans):
    if len(stack) == len(nums):
      ans.append(stack[:])
      return

    for i in range(len(nums)):
      # same value should appear in index order
      if used[i] or i > 0 and nums[i] == nums[i - 1] and not used[i - 1]:
        continue
      used[i] = True
      stack.append(nums[i])
      dfs(nums, stack, used, ans)
      stack.pop()
      used[i] = False 

  nums.sort()
  ans = []
  dfs(nums, [], [False] * len(nums), ans)
  return ans


print(permuteUnique([1, 1, 2]))
print(permuteUnique([0]))
print(permuteUnique([1, 1, 1, 1]))
print(permuteUnique([0, 1, 1, 1, 0]))

File: competitive_programming/leetcode/cpp/47_permutations_ii.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void dfs(int used, vector<int>& stack, vector<vector<int>>& ans, const vector<int>& nums) {
    if (stack.size() == nums.size()) {
        ans.push_back(stack);
        return;
    }
    for (int i = 0; i < nums.size(); ++i) {
        if (i > 0 && nums[i] == nums[i - 1] && !(used & 1 << (i - 1))) {
            continue;
        }
        if (!(used & 1 << i)) {
            stack.push_back(nums[i]);
            dfs(used | 1 << i, stack, ans, nums);
            stack.pop_back();
        }
    }
}

vector<vector<int>> permuteUnique(vector<int>& nums) {
    vector<vector<int>> ans;
    vector<int> stack;
    sort(nums.begin(), nums.end());
    dfs(0, stack, ans, nums);
    return ans;
}

void print(vector<vector<int>> ans) {
    for (vector<int> arr : ans) {
        for (int i : arr) {
            cout << i << " ";
        }
        cout << "\n";
    }
}

int main() {
    vector<int> nums = {1, 1, 3};
    vector<vector<int>> ans = permuteUnique(nums);
    print(ans);
}